### PR TITLE
M-LM - M-Bridge Data consolidation

### PR DESCRIPTION
**Pull request by asolergi-nv**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [1e1df9c](https://github.com/NVIDIA/Megatron-LM/pull/5409/commits/1e1df9c4f748ec73a69f9ac49a51271c96a98a9c)

**Comment by asolergi-nv** [comment]
/ok to test [c583d71](https://github.com/NVIDIA/Megatron-LM/pull/5409/commits/c583d71d9a2fe02d1c7dde287363b8e30d75cedf)

**Comment by asolergi-nv** [comment]
/ok to test [0d77126](https://github.com/NVIDIA/Megatron-LM/pull/5409/commits/0d771260fa280b6c6895a72f65f56e040a02d0e6)

File: megatron/core/datasets/blended_megatron_dataset_config.py [hunk 1]
```diff
@@ -96,8 +96,35 @@ class BlendedMegatronDatasetConfig:
...
+
+    persistent_workers: bool = True
+    """Whether the torch DataLoader keeps worker subprocesses alive between epochs.
+       Forced to False when num_workers == 0.
+    """
+
     def __post_init__(self) -> None:
         """Do asserts and set fields post init"""
+        # Persistent workers require at least one worker subprocess.
+        if self.num_workers == 0 and self.persistent_workers:
```
Comment (maanug-nv) [inline megatron/core/datasets/blended_megatron_dataset_config.py:126]: nit: Should probably log a warning for this.

File: megatron/training/training.py [hunk 2]
```diff
@@ -4244,8 +4224,14 @@ def build_train_valid_test_datasets(build_train_valid_test_datasets_provider, tr
     return build_train_valid_test_datasets_provider(train_valid_test_num_samples)
 
 
-def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider):
-    """Build pretraining data loaders."""
+def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider, cfg=None):
+    """Build pretraining data loaders.
+
+    When ``cfg`` (a ``PretrainConfigContainer`` with a populated ``cfg.dataset``)
```
Comment (maanug-nv) [inline megatron/training/training.py:4230]: nit: seems like this docstring should belong in `build_train_valid_test_data_iterators()`

File: megatron/training/config/container.py [hunk 3]
```diff
@@ -236,7 +237,7 @@ class PretrainConfigContainer(ConfigContainerBase):
     model: HybridModelConfig | GPTModelConfig
     optimizer: OptimizerConfig
     scheduler: SchedulerConfig
-    # dataset: GPTDatasetConfig # TODO (@maanug): add support
+    dataset: Optional[BlendedMegatronDatasetConfig] = None
```
Comment (maanug-nv) [inline megatron/training/config/container.py:240]: i understand why this is marked as optional right now, but conceptually this isn't really optional for pretraining. what are the blockers/challenges for ensuring we can always provide a dataset config to the config container?

File: megatron/training/datasets/data_loaders.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+            return RerunDataIterator(dataloader)
+        else:
+            raise RuntimeError("unexpected dataloader type")
+
+    # GPT datasets always cycle validation so eval_iters may exceed valid size.
+    is_gpt_dataset = _is_gpt_dataset_config(cfg.dataset)
+
+    train_it = _get_iterator(dl_type, train_dataloader) if train_dataloader is not None else None
+    if valid_dataloader is not None:
+        valid_it = _get_iterator("cyclic" if is_gpt_dataset else dl_type, valid_dataloader)
```
Comment (maanug-nv) [inline megatron/training/datasets/data_loaders.py:433]: seems like this is missing support for multiple validation sets

File: megatron/training/datasets/data_loaders.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    else:
+        valid_it = None
+    test_it = _get_iterator(dl_type, test_dataloader) if test_dataloader is not None else None
+
+    return train_it, valid_it, test_it
+
+
+def build_train_valid_test_data_iterators(
```
Comment (maanug-nv) [inline megatron/training/datasets/data_loaders.py:441]: seems like this is not used anywhere? shouldn't it replace the one in `training/training.py`?

File: megatron/training/datasets/data_samplers.py [hunk 6]
```diff
@@ -94,22 +100,60 @@ def close_nvidia_fds():
...
-        worker_init_fn if args.num_workers > 0 else None
+    maybe_worker_init_fn = worker_init_fn if num_workers > 0 else None
+
+    # Megatron-LM-specific samplers not present in the consolidated core: build
+    # the DataLoader here. (full_validation / hybrid_context_parallel are global
+    # parallelism settings, so they are read from args regardless of source.)
+    batch_sampler = None
+    extra_kwargs = {}
+    if split == Split.valid and args.full_validation:
+        batch_sampler = MegatronFullValidationSampler(
```
Comment (maanug-nv) [inline megatron/training/datasets/data_samplers.py:111]: why are these samplers only supported here, and not in the `data_loaders.py` version?

**Review by maanug-nv — COMMENTED** [review]
A few concerns. Mainly, i don't understand why functions in `data_loaders.py` arent replacing the ones in `training/training.py` of the same name. also, the same functions seem to be missing some features like multiple validation sets.

File: megatron/training/datasets/data_samplers.py (hunk 6 above)
Comment (maanug-nv) [reply megatron/training/datasets/data_samplers.py:111]: to expound on this concern, if someone adds a new sampler to this path (data_samplers.build_pretraining_data_loader) and then we later need support for that in Bridge, what do we do? we'd likely need to ask the author of the new sampler to refactor.
